DVCodingStyle.md

```diff
@@ -454,13 +455,194 @@ utility macros for even more common code sequences.
+        file.
+
+        :-1:
+        ```systemverilog
+        // file: dv_macros.h
+        automatic function void get_nco();
+          ...
+        endfunction
+        // Bad: Putting functions and tasks in a header file that is potentially
+        //      included multiple times.
+
+        `define DV_FOO(...) ...
+        // Bad: No ifndef guard. Macro re-definition warnings will be thrown if
+        //      this header file is included in multiple sources.
+        ```
+
+        :+1:
+        ```systemverilog
+        // file: dv_macros.svh
+        `ifndef DV_FOO
+          `define DV_FOO(...) ...
+        `endif
+
+        `ifndef DV_BAR
+          `define DV_BAR xyz
+        `endif
+        ```
+
+        Generally, there is no need to `` `include`` this header file if it
+        can be guaranteed that it is already compiled prior to the current
+        compilation unit in which those macros are invoked. This is typically
+        handled by an external tool that works out the dependencies across
+        the compilation units (example: FuseSoC). However, it may not always be
+        possible to guarantee this, and the `` `include`` may be necessary.
+        The `` `ifndef`` guard helps prevent the re-definition of the macros with
+        itself.
+
+
+    *   If the scope of the macros defined is limited to a package or source
+        (individual interface, module or a class), then place them at the top of
+        the file. Always `` `undef`` them at the end of the file so that they
+        are no longer visible to subsequent sources during compilation. Do not
+        put `` `ifndef`` guards on these macros, so that the re-definitions
+        if encountered, are flagged.
+
+        :-1:
+        ```systemverilog
+        // file: foo_init_seq.sv
+        `ifndef GET_DATA
+          `define GET_DATA(baz) ...
+        `endif
+        // Bad: Macro-redefinition if encountered, will be ignored due to the
+        //      ifndef guard. Previously defined macro is invoked instead.
+        // Bad: Macro name is not specific enough.
+
+        class foo_init_seq;
+          ...
+        endclass
+        // EOF
+        // Bad: Macro is not undefined at the end of the file.
+        ```
+
+        :+1:
+        ```systemverilog
+        // file: foo_init_seq.sv
+        `define GET_FOO_TX_FIFO_DATA_AFTER_OP(baz) ...
+
+        class foo_init_seq;
+          ...
+        endclass
+
+        `undef GET_FOO_TX_FIFO_DATA_AFTER_OP
+        ```
+
+        :+1:
+        ```systemverilog
+        // file: foo_env_pkg.sv
+        `define GET_FOO_TX_FIFO_DATA(baz) ...
+
+        ...
+
+        // packages sources
+        `include "foo_env_cfg.sv"
+        ...
+        `include "foo_scoreboard.sv"
+        `include "foo_env.sv"
+
+
+        `undef GET_FOO_TX_FIFO_DATA
+        // Good: All package sources can invoke this macro. It will no longer
+        //       be visible beyond this compilation unit.
```

Comment:
Does SystemVerilog have a notion of "compilation unit" like C and similar languages? Surely, the whole pain of undefining macros is because it doesn't?

Also, I think it might be clearer to read this if the two halves of the sentence were joined. So something like this, maybe? "All package sources can invoke this macro, but it is not visible in any other code".

Reply:
Reworded a bit.